src/editor/Editor.js

```diff
@@ -2385,6 +2401,128 @@ define(function (require, exports, module) {
+        }
+
+        function _sortByPriority(a, b) {
+            return a.priority - b.priority;
+        }
+
+        function _getName(gutter) {
+            return gutter.name;
+        }
+
+        var gutters = this._registeredGutters.map(_getName),
+            rootElement = this.getRootElement();
+
+        // If the line numbers gutter has not been explicitly registered and the CodeMirror lineNumbes option is
+        // set to true, we explicitly add the line  numbers gutter. This case occurs the first time the editor loads
+        // and showwLineNumbers is set to true in preferences
+        if (gutters.indexOf(LINE_NUMBER_GUTTER) < 0 && this._codeMirror.getOption(cmOptions[SHOW_LINE_NUMBERS])) {
+            this._registeredGutters.push({name: LINE_NUMBER_GUTTER, priority: 100});
+        }
+
+        gutters = this._registeredGutters.sort(_sortByPriority)
+            .filter(_filterByLanguages)
+            .map(_getName);
+
+        this._codeMirror.setOption("gutters", gutters);
+
+        this._codeMirror.refresh();
+
+        if (gutters.indexOf(LINE_NUMBER_GUTTER) < 0) {
+            $(rootElement).addClass("linenumber-disabled");
+        } else {
+            $(rootElement).removeClass("linenumber-disabled");
+        }
+    };
+
+    /**
+     * Sets the marker for the specified gutter on the specified line number
+     * @param   {string}   lineNumber The line number for the inserted gutter marker
+     * @param   {string}   gutterName The name of the gutter
+     * @param   {object}   marker     The jquery object representint the marker to the inserted in the gutter
+     */
+    Editor.prototype.setGutterMarker = function (lineNumber, gutterName, marker) {
+        var gutterNameRegistered = this._registeredGutters.some(function (gutter) {
+            return gutter.name === gutterName;
+        });
+
+        if (!gutterNameRegistered) {
+            console.warn("Gutter name must be registered before calling Editor.setGutterMarker");
+            return;
+        }
+
+        this._codeMirror.setGutterMarker(lineNumber, gutterName, marker);
+    };
+
+    /**
+     * Returns the list of gutters current registered on this editor.
+     * @return {!Array.<{name: string, priority: number}>}
+     */
+    Editor.prototype.getRegisteredGutters = function () {
+        return this._registeredGutters;
+    };
+
+    /**
+     * Registers the gutter with the specified name at the given priority.
+     * @param {string} name    The name to register with the quarter
+     * @param {number} priority  A number denoting the priority of the gutter. Priorities higher than 100 appear after the line numbers. Priority less than 100 appear before.
+     * @param {?Array<number>} languageIds A list of language ids that this gutter is valid for. If no language ids are passed, then the gutter is valid in all languages.
+     */
+    Editor.prototype.registerGutter = function (name, priority, languageIds) {
```

Comment:
correct me if I'm wrong, but every editor instance is language specific and gutters live in it's instance property `_registeredGutters` so how does language make sense here? can language of the editor instance change during its lifetime?